lib/component-data/actions.js

```diff
@@ -357,38 +357,55 @@ function addComponentsToPageArea(store, {currentURI, path, replace, components})
-      // because they exist in page data
-      return Promise.all(_.map(newComponents, (component) => {
-        const uri = component[refProp];
-
-        delete component[refProp]; // remove uri from actual data, since we're not doing a cascading PUT
-        return modelSave(uri, component)
-          .then((savedData) => addToQueue(save, [uri, savedData, false], 'save'))
-          .then((savedData) => publish(uri, savedData, null, false, store))
-          .then((savedData) => modelRender(uri, savedData))
-          .then((savedData) => store.commit(UPDATE_COMPONENT, {uri, data: savedData}))
-          .then(() => uri);
-      }))
-        .then((uris) => {
-          if (replace) {
-            // replace current uri when creating new list
-            return _.slice(data, 0, index).concat(uris).concat(_.slice(data, index + 1));
-          } else {
-            // just add new components after the index
-            return _.slice(data, 0, index + 1).concat(uris).concat(_.slice(data, index + 1));
-          }
-        })
-        .then((newList) => {
-          if (replace) {
-            store.commit(REMOVE_COMPONENT, { uri: currentURI });
-          }
-          // save page
-          return store.dispatch('savePage', { [path]: newList });
-        });
+      store.dispatch('addCreatedComponentsToPageArea', {newComponents, store, currentURI, path, replace, index, data});
+    });
+}
+
+/**
+ * add components to a page area
+ * @param {object} store
+ * @param {array} newComponents
+ * @param {string} currentURI
+ * @param {string} path
+ * @param {boolean} replace
+ * @param {index} number
+ * @param {data} array
+ * @param {forceRender} boolean
+ * @returns {Promise}
+ */
+export function addCreatedComponentsToPageArea(store, { newComponents,  currentURI, path, replace, index, data, forceRender}) {
+  // note: we have to explicitly save these (rather than relying on the cascading PUT of the parent),
+  // because they exist in page data
+  return Promise.all(_.map(newComponents, (component) => {
+    const uri = component[refProp] || currentURI;
+
+    delete component[refProp]; // remove uri from actual data, since we're not doing a cascading PUT
+
+    return modelSave(uri, component)
+      .then((savedData) => addToQueue(save, [uri, savedData, false], 'save'))
+      .then((savedData) => publish(uri, savedData, null, false, store))
+      .then((savedData) => modelRender(uri, savedData))
+      .then((savedData) => store.commit(UPDATE_COMPONENT, {uri, data: savedData}))
+      .then(() => uri);
+  })).then((uris) => {
+    if (replace) {
+      // replace current uri when creating new list
+      return _.slice(data, 0, index).concat(uris).concat(_.slice(data, index + 1));
+    } else {
+      // just add new components after the index
+      return _.slice(data, 0, index + 1).concat(uris).concat(_.slice(data, index + 1));
+    }
```

Comment:
To reduce a couple lines of codes, we could use ternary operator to store the value of the `index` (since that's the only value that changes from the statement) and have a single `return` statement. Here's a quick example
```js
const seletcedIndex = replace ? index : index + 1;

return _.slice(data, 0, selectedIndex).concat(uris).concat(_.slice(data, index + 1));
```

Reply:
I switched to the ternary, although this wasn't actually code that I changed.  It was already like that.  I just broke one function into two, so it looks as though I added it.